### DISCUSSION TITLE
`KeyError` when trying to remove rule from PhraseMatcher using match ID

### DISCUSSION THREAD

**Discussion by raqibhayder** [explosion/spaCy#9874] [Help: Other Questions] [answered]
## How to reproduce the behaviour
Unable to reproduce the behaviour. It happens occasionally in the production environment. 

## Details
I have a custom component called `hints` which I use to match entities before passing it to the NER component in the pipeline. From time to time I get `KeyError` when trying to remove the `match ID` from the PhraseMatcher instance. I know I can fix this by checking if the `match ID` exists before removing but I want to understand the reason for such behaviour. 

For every call, I add the pattern with the associated `match ID`. For example like so:
``` python
self._phrase_matcher.add(Hints.PERSON.value, [person_hint_pattern])
```

Therefore, I should be able to remove the pattern based on `match ID` too. For example like so:
``` python
self._phrase_matcher.remove(Hints.PERSON.value)
```

What am I missing?

Thank you for you help 🙏🏽 

---
``` python
@Language.factory("hints")
def hints_component(nlp, name):
    logger.info(f"Adding {name} component.")
    return HintsComponent(nlp=nlp)


class Hints(str, Enum):
    PERSON = "PERSON"
    ORG = "ORG"


class HintsComponent:
    """HintsComponent uses rule-based phrase matching to tag entities.

    Hints are added to the Doc object post tokenization. The HintsComponent:
    1. Accesses the hints using doc._.hints.
    2. Adds hints as patterns to the PhraseMatcher.
    3. If matches are found, use the matches to create entity spans and assign
       entity labels to the spans based on the hint type.
    4. Removes the patterns (hints) from the PhraseMatcher to ensure subsequent
       requests do not use stale hints.
    """

    # Add hints extension to Doc object. doc._.hints is set post tokenization
    if not Doc.has_extension("hints"):
        Doc.set_extension("hints", default={})

    def __init__(self, nlp: Language):
        """Initialize the HintsComponent.

        Args:
            nlp: spaCy language model
        """

        self._nlp = nlp
        self._phrase_matcher = PhraseMatcher(nlp.vocab,
                                             attr="TEXT",
                                             validate=True)

    def __call__(self, doc) -> Doc:
        if not doc._.hints:
            return doc

        hints = doc._.hints

        person_hint_pattern = self._nlp(hints[Hints.PERSON])
        org_hint_pattern = self._nlp(hints[Hints.ORG])

        self._phrase_matcher.add(Hints.PERSON.value, [person_hint_pattern])
        self._phrase_matcher.add(Hints.ORG.value, [org_hint_pattern])

        matches = self._phrase_matcher(doc, as_spans=True)
        # In case of overlapping matches, keep the longest match
        doc.ents = spacy.util.filter_spans(matches)

        with doc.retokenize() as retokenizer:
            for ent in doc.ents:
                retokenizer.merge(ent)

        # Remove patterns to ensure that hints from previous requests are not used
        # in successive requests
        self._phrase_matcher.remove(Hints.PERSON.value)
        self._phrase_matcher.remove(Hints.ORG.value)


        return doc
```

The component is used like so:
``` python
import spacy 
nlp = spacy.load("en_core_web_md")
nlp.add_pipe(factory_name="hints", name="hints", before="ner")
doc = nlp.make_doc("John Smith works at ACME")
doc._.hints= {"PERSON":"John Smith",  "ORG":"ACME"}
doc = nlp(doc)
```

## Your Environment

* Operating System: Debian
* Python Version Used: 3.9.7
* spaCy Version Used: 3.2.1

**Comment by adrianeboyd** [comment]
How are you using this in practice? Are you only running it from a simple `python script.py` as in the example?

If you initialize a new phrase matcher in each `__call__` instead of `__init__` (so no need to use `remove`) does this solve the problem?

```python
        self._phrase_matcher = PhraseMatcher(nlp.vocab,
                                             attr="TEXT",
                                             validate=True)
```

The phrase matcher init should not be expensive.

**Comment by raqibhayder** [comment]
> How are you using this in practice? Are you only running it from a simple python script.py as in the example?

No. I am running this as part of a FastAPI application (with Gunicorn and UvicornWorkers). I load the `en_core_web_md` and add the component to the pipeline like above. 

> If you initialize a new phrase matcher in each __call__ instead of __init__ (so no need to use remove) does this solve the problem?

I should have done that but was worried about how expensive it would be. I will change that and let you know how it goes. 


As always, Thank you for always replying so promptly. 🙏🏽

**Comment by adrianeboyd** [comment, marked as answer]
Using FastAPI would explain why you were running into problems, since you can end up with a configuration where multiple threads are trying to modify the PhraseMatcher object at the same time. Some relevant FastAPI docs: https://fastapi.tiangolo.com/async/

My guess would be that creating a new phrase matcher each time is not going to be much slower (if at all) than trying to do the remove.

**Comment by raqibhayder** [comment]
> My guess would be that creating a new phrase matcher each time is not going to be much slower (if at all) than trying to do the remove.

You are right. Creating new instances of PhraserMatcher every call has negligible overhead. Also, using `Language.make_doc` speeds everything up significantly. 

Thank you @adrianeboyd
